include/cachemere/cache.hpp

```diff
@@ -247,163 +294,156 @@ std::unique_lock<std::recursive_mutex> Cache<K, V, I, E, SV, SK, TS>::lock() con
-        key_and_item->second.m_total_size = key_size + value_size;
-
-        on_update(key_and_item->first, key_and_item->second);
-    } else {
-        // Insert.
-        const auto it_and_ok =
-            m_data.emplace(std::piecewise_construct, std::forward_as_tuple(std::move(key)), std::forward_as_tuple(key_size, std::move(value), value_size));
-        assert(it_and_ok.second);
-
-        on_insert(it_and_ok.first->first, it_and_ok.first->second);
-
-        m_current_size += key_size + key_size + value_size;
+    if (m_constraint_policy->can_replace(key, old_item, new_item)) {
+        return true;
     }
-}
 
-template<class K, class V, template<class, class> class I, template<class, class> class E, class SV, class SK, bool TS>
-size_t Cache<K, V, I, E, SV, SK, TS>::free_amount(size_t amount_to_free)
-{
-    assert(amount_to_free <= m_current_size);
-    size_t freed_amount = 0;
+    // The logic here is very similar to check_insert but with an important modification.
+    // Since in this code path we're updating an existing key instead of adding a new one, we need to handle the case
+    // where the eviction policy recommends the eviction of the key we're trying to insert. If this happens and the constraint
+    // is still not satisfied afterwards, we need to treat all subsequent constraint checks as if we were inserting instead of updating our key
+    // (because the original was evicted).
+    auto constraint_copy      = std::make_unique<MyConstraintPolicy>(*m_constraint_policy);
+    bool evicted_original_key = false;
+
+    auto can_replace = [&]() {
+        if (evicted_original_key) {
+            return constraint_copy->can_add(key, new_item);
+        } else {
+            return constraint_copy->can_replace(key, old_item, new_item);
+        }
+    };
 
-    auto should_evict_another = [&](auto victim_it) { return victim_it != m_eviction_policy->victim_end() && freed_amount <= amount_to_free; };
+    auto should_evict_another = [&](auto victim_it) { return victim_it != m_eviction_policy->victim_end() && !can_replace(); };
 
-    for (auto victim_it = m_eviction_policy->victim_begin(); should_evict_another(victim_it); victim_it = m_eviction_policy->victim_begin()) {
+    // Evict until the constraint copy is happy.
+    std::vector<DataMapIt> keys_to_evict;
+
+    for (auto victim_it = m_eviction_policy->victim_begin(); should_evict_another(victim_it); ++victim_it) {
         const K& key_to_evict = *victim_it;
         auto     key_and_item = m_data.find(key_to_evict);
 
         if (key_and_item != m_data.end()) {
-            auto total_size_and_overhead = key_and_item->second.m_key_size + key_and_item->second.m_total_size;
-            freed_amount += total_size_and_overhead;
-            remove(key_and_item);
+            if (!m_insertion_policy->should_replace(key_to_evict, key)) {
+                return false;
+            }
+
+            evicted_original_key |= key_and_item->first == key;
```

Comment:
This is technically wasteful. You're going to compare all keys you iterate over even if you already found the original key, since this is a bitwise or and there's no short-circuiting.
```suggestion
            if (!evicted_original_key) {
                evicted_original_key = key_and_item->first == key;
            }
```